modules/contact-form/grunion-contact-form.php

```diff
@@ -844,16 +844,43 @@ public function personal_data_exporter( $email, $page = 1 ) {
 	 */
 	public function personal_data_eraser( $email, $page = 1 ) {
 		$per_page = 250;
-		$removed  = 0;
-		$retained = 0;
+		$removed  = false;
+		$retained = false;
 		$messages = array();
 		$post_ids = $this->personal_data_post_ids_by_email( $email, $per_page, $page );
```

Comment:
@coreymckrill I was looking into this implementation when adding it to WPJM and I noticed what I think is a bug here. I think the pagination will break this because for each pass, (ideally) a page worth of posts will be removed. The new query will run but with the assumption that those posts are there. For WPJM, I think I might get around this by trashing the posts instead of deleting them and including that post status in my query. I could reset to page 1 for each pass, but then if a post isn't deleted (due to result of filter or failed deletion) the messages will be duplicated for every run. Also, with the `page=1` option, if a page worth of posts aren't deleted due to this new filter, then I think it will just keep running.

Reply:
@jom The `$page` param is[ enumerated by core](https://github.com/WordPress/wordpress-develop/blob/8adbf40bc260e209199cb963cb0cd62a5873e125/src/wp-admin/includes/ajax-actions.php#L4624) each time the eraser is called. Doesn't that allow these to be paginated in full (i.e., not run forever) even when some are not erased from the database?

Reply:
The issue of it possibly running forever would only happen if you always pass `1` for the `$page` parameter to the `personal_data_post_ids_by_email()` method (as one fix to the issue happening here). The main issue here is that, as is, the page parameter will skip deleting pages of form responses. If there are, for example, 550 contact form responses (haha) for an email address, I think this will currently run:
page 1: `LIMIT 0, 250`, immediately delete 250 (1-250) responses.
page 2: `LIMIT 250, 250`, deleting responses originally marked 500-550 and skipping the next batch of 250 (originally results 251-500) because the first batch was deleted. `done` will pass `true` here even though there are 250 responses yet to delete.
page 3 would never run, but if it did: `LIMIT 500, 250` will return an empty set as it offsets past the 250 that were skipped.

Leaving 250 responses.

Reply:
That sounds plausible, though I haven't tested it. One way to get around it might be to take the same route that Core does with comments and [anonymize the data](https://github.com/WordPress/wordpress-develop/blob/8adbf40bc260e209199cb963cb0cd62a5873e125/src/wp-includes/comment.php#L3401-L3500) instead of deleting the posts completely.